### PR TITLE
SD-4070 HTTP Push - send published package and its items in a single request in specified format

### PR DESCRIPTION
**Pull request by mugurrus**
Embed package items in the sent package for HTTP push publish
destination.

### CODE REVIEW

File: apps/packages/package_service.py [hunk 1]
```diff
@@ -300,8 +301,10 @@ def replace_ref_in_package(self, package, old_ref_id, new_ref_id):
         non_root_groups = (group for group in package.get(GROUPS, []) if group.get(GROUP_ID) != ROOT_GROUP)
         for g in (ref for group in non_root_groups for ref in group.get(REFS, [])):
             if g.get(RESIDREF, '') == old_ref_id:
+                new_item = get_resource_service('archive').find_one(req=None, _id=new_ref_id)
                 g[RESIDREF] = new_ref_id
                 g['guid'] = new_ref_id
+                g[VERSION] = new_item[VERSION]
```
Comment (akintolga) [inline apps/packages/package_service.py]: Do we need to save the version of items in a unpublished package? Until it gets published a package always points to the last version of the referenced item.

File: apps/publish/enqueue/enqueue_service.py [hunk 2]
```diff
@@ -277,6 +283,24 @@ def queue_transmission(self, doc, subscribers):
...
+        """ Embeds all package items in the package document
+        """
+        for group in package.get(GROUPS, []):
+            if group[GROUP_ID] == ROOT_GROUP:
+                continue
+            for ref in group[REFS]:
+                if RESIDREF not in ref:
+                    continue
+                package_item = get_resource_service('published').find_one(req=None, item_id=ref['residRef'],
+                                                                          _current_version=ref['_current_version'])
```
Comment (akintolga) [inline apps/publish/enqueue/enqueue_service.py]: VERSION enum would be good here...

File: apps/publish/enqueue/enqueue_service.py [hunk 3]
```diff
@@ -266,8 +270,10 @@ def queue_transmission(self, doc, subscribers):
                             publish_queue_item['content_type'] = doc.get('type', None)
                             publish_queue_item['headline'] = doc.get('headline', None)
                             publish_queue_item['publishing_action'] = self.published_state
+                            if doc.get(PUBLISHED_IN_PACKAGE):
+                                publish_queue_item[PUBLISHED_IN_PACKAGE] = doc[PUBLISHED_IN_PACKAGE]
                             publish_queue_item.pop(ITEM_STATE, None)
-                            get_resource_service('publish_queue').post([publish_queue_item])
+                            get_resource_service('publish_queue').post([publish_queue_item])[0]
```
Comment (akintolga) [inline apps/publish/enqueue/enqueue_service.py]: Are we doing anything with the first item in this line?

File: apps/publish/content/common.py [hunk 4]
```diff
@@ -131,8 +133,8 @@ def update(self, id, updates, original):
                 self._update_archive(original, updated, should_insert_into_versions=auto_publish)
                 self.update_published_collection(published_item_id=original[config.ID_FIELD], updated=updated)
 
-                from apps.publish.enqueue import enqueue_published
-                enqueue_published.apply_async()
+            from apps.publish.enqueue import EnqueueContent
+            EnqueueContent().run()
```
Comment (mdhaman) [inline apps/publish/content/common.py]: why synchronous queuing?

File: apps/publish/enqueue/enqueue_service.py (hunk 2 above)
Comment (sivakuna-aap) [inline apps/publish/enqueue/enqueue_service.py]: String constant for `residRef` also

File: apps/publish/enqueue/enqueue_service.py (hunk 3 above)
Comment (mugurrus) [inline apps/publish/enqueue/enqueue_service.py]: Will remove, used it to debug

File: apps/packages/package_service.py (hunk 1 above)
Comment (mugurrus) [inline apps/packages/package_service.py]: When embedding items in a package they are read from published collection. In this collection there can be multiple records with the same item id so I need to be able to select the one that corresponds to the published package.

**Comment by ioanpocol** [comment]
:+1:

**Comment by petrjasek** [comment]
would this version work for you @akintolga @mdhaman or is this something that should be configurable?

**Comment by petrjasek** [comment]
my bad, config is there. :+1:

**Comment by akintolga** [comment]
:+1:

**Pull request merged**